### PR TITLE
Lab review 01

### PR DESCRIPTION
**Pull request by u-fischer**
This adapts a few of the latex-lab files to the new hooks with arguments and corrects a number of small bugs. The bib-module handles now biblatex if hyperref is loaded too. The tagpdf version has been updated (which corrects a bug with empty structures). A new firstaid module collects some changes to external packages and classes , it is not loaded automatically by phase-III.

# Internal housekeeping

## Status of pull request

- Ready to merge

## Checklist of required changes before merge will be approved
- [ x] Test file(s) added (tests for changes in external packages must be called locally)
- [ x] Version and date string updated in changed source files
- [ ] Relevant `\changes` entries in source included
- [ x] Relevant `changes.txt` updated
- [ ] `ltnewsX.tex` (and/or `latexchanges.tex`) updated  (not sure if the firstaid should be mentioned?)

### CODE REVIEW

**Review by FrankMittelbach — COMMENTED** [review]
looked it over only up firstaid, but have something coming up for the next hours

File: required/latex-lab/latex-lab-testphase.dtx
Comment (FrankMittelbach) [inline required/latex-lab/latex-lab-testphase.dtx:88]: thought we wanted to drop -tagging (not important though)

File: required/latex-lab/latex-lab-toc-kernel-changes.dtx
Comment (FrankMittelbach) [inline required/latex-lab/latex-lab-toc-kernel-changes.dtx:187]: we should review if we really want to turn individual generic hooks into hooks with arguments. I think we should not and instead use \AddTohookWithArguments{addcontentsline} here and leave the generic one for code that goes around the whole command for one or the other reason. It is not less efficient (ie when the generic hook is not filled it is not patched in)

File: required/latex-lab/latex-lab-toc-kernel-changes.dtx
Comment (FrankMittelbach) [inline required/latex-lab/latex-lab-toc-kernel-changes.dtx:238]: TODO: convince myself if we really need all of these hooks, I doubt it. In particular I doubt that the /after hooks need to have arguments (or need to be there).

**Review by FrankMittelbach — APPROVED** [review]
I may be wrong with some of the remarks, but I think a few of the should be addressed now  (or refuted) others could be addressed later after a merge.

File: required/latex-lab/latex-lab-block-tagging.dtx
Comment (u-fischer) [reply required/latex-lab/latex-lab-block-tagging.dtx:1350]: yes it isn't anymore (the text was added before). But imho `\bfseries` is better nevertheless, in case some document adds tagging to `\textbf`. I changed the text a bit.

File: required/latex-lab/latex-lab-testphase.dtx
Comment (u-fischer) [reply required/latex-lab/latex-lab-testphase.dtx:88]: yes, next round. I did run out of time  ;-)

File: required/latex-lab/latex-lab-toc-kernel-changes.dtx
Comment (u-fischer) [reply required/latex-lab/latex-lab-toc-kernel-changes.dtx:238]: well I need to inject link + tagging code around the number, the text and the page. And this code must be mirrored: the link code must be either inside or outside the tagging code. I don't quite see how  do it differently or how to produce a pair where the begin takes arguments and the end hook not.

Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-toc-kernel-changes.dtx:236]: make the number an argument to the hook maybe, really comes down to configuration points by the end of the day ... but it isn't something that need to be solved or changed now, but perhaps add a TODO to review the current approach

File: required/latex-lab/config-firstaid.lua
Comment (FrankMittelbach) [reply required/latex-lab/config-firstaid.lua:4]: my position is that the tests for firstaid help to notice that something got corrected and the firstaid is no longer nevessary (or even wrong). And once, the firstaid is no longer required the corresponding test can go again. But yes these are judgement calls

File: required/latex-lab/latex-lab-bib.dtx
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-bib.dtx:179]: ok

File: required/latex-lab/latex-lab-block-tagging.dtx
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-block-tagging.dtx]: yes \bfseries is better (in fact \textbf is probably wrong as the user input may have or wants \par inside the caption)

File: required/latex-lab/latex-lab-toc-kernel-changes.dtx
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-toc-kernel-changes.dtx]: as I said: I think we should not  change generic hooks to become hooks with arguments. I think this leads to bad style and difficulties. Why is the generic hook filled? It is only filled by the tagging code so if that goes into a hook by itself then the generic hook would normally not be there and would be free to be used on the outside for somehting else if needed.

**Comment by josephwright** [comment]
@FrankMittelbach  Are we good to go here?

**Comment by FrankMittelbach** [comment]
> @FrankMittelbach Are we good to go here?

I would suggest not to make generic hooks hooks with arguments (I think just one case), otherwise yes.

File: required/latex-lab/latex-lab-toc-kernel-changes.dtx
Comment (u-fischer) [reply required/latex-lab/latex-lab-toc-kernel-changes.dtx]: I already removed the declaration and the UseHook command hook.  (But I don't quite understand the distinction, we just made generic hooks into hooks with argument).

**Comment by FrankMittelbach** [comment]
> > > @FrankMittelbach Are we good to go here?
> > 
> > 
> > I would suggest not to make generic hooks hooks with arguments (I think just one case), otherwise yes.
> 
> OK, I'll wait for that to be changed

I meant use them in that case, but as I said that was partly nonsense

**Comment by josephwright** [comment]
Merged, letting this build then will start on the release (will be tomorrow now)

**Pull request merged**
